takashi-market-maker.md

```diff
@@ -0,0 +1,170 @@
+# Market Maker
+
+## Algorithm
+
+> Write some code to create an algorithm that would maximize the amount of
+> money, assuming that nobody else buys and sells.
+
+some pseudocode:
+
+```py
+for order in orders:
+  curr_asset_price = lookup_asset_price(order[asset])
+  if order[buy] and order[price] >= curr_asset_price * 1.01:
+    execute_sell_order(order)
+  elif order[sell] and order[price] <= curr_asset_price * .99:
+    execute_buy_order(order)
+```
+
+> What happens if this becomes an n-sided market, where there are many
+> market makers? How does this change the problem?
+
+If this becomes a n-sided problem, speed matters quite a lot -- if you
+wait, someone else will snatch up the trade. Also, with more
+competition, the spread between bid and ask might have to tighten, but
+if there's less competition, the bid and ask spread can widen. You can
+code up an algorithm that adjusts automatically to market volume in
+this case.
```

Comment:
how would you know what an appropriate threshold is? would it vary from stock to stock? if so do we need to have some sort of DB storing historical profits for each trade

Reply:
So when ingesting data from a data source with financial data, you get a bunch of other data as well:

Take this curl command, to grab data from yahoo finance for AAPL:
```sh
curl 'https://query1.finance.yahoo.com/v8/finance/chart/AAPL?region=US&lang=en-US&includePrePost=false&interval=15m&useYfid=true&range=5d&corsDomain=finance.yahoo.com&.tsrc=finance' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.2 Safari/605.1.15' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br' -H 'Referer: https://finance.yahoo.com/quote/AAPL/' -H 'Origin: https://finance.yahoo.com' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Cookie: A1=d=AQABBOpSI2ICEGjjE16Xa3Y8E2kAx-FYilQFEgEBAQGkJGItYgAAAAAA_eMAAAcI6lIjYuFYilQ&S=AQAAAnVw8S45YkuucaDtlKlUp2M; A3=d=AQABBOpSI2ICEGjjE16Xa3Y8E2kAx-FYilQFEgEBAQGkJGItYgAAAAAA_eMAAAcI6lIjYuFYilQ&S=AQAAAnVw8S45YkuucaDtlKlUp2M; A1S=d=AQABBOpSI2ICEGjjE16Xa3Y8E2kAx-FYilQFEgEBAQGkJGItYgAAAAAA_eMAAAcI6lIjYuFYilQ&S=AQAAAnVw8S45YkuucaDtlKlUp2M&j=US; B=592ios5h26kna&b=3&s=q3; GUC=AQEBAQFiJKRiLUIbXAQr; PRF=t%3DAAPL' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-site'
```

This returns a schema that looks like this:

```go
type AutoGenerated struct {
	Chart struct {
		Result []struct {
			Meta struct {
				Currency             string  `json:"currency"`
				Symbol               string  `json:"symbol"`
				ExchangeName         string  `json:"exchangeName"`
				InstrumentType       string  `json:"instrumentType"`
				FirstTradeDate       int     `json:"firstTradeDate"`
				RegularMarketTime    int     `json:"regularMarketTime"`
				Gmtoffset            int     `json:"gmtoffset"`
				Timezone             string  `json:"timezone"`
				ExchangeTimezoneName string  `json:"exchangeTimezoneName"`
				RegularMarketPrice   float64 `json:"regularMarketPrice"`
				ChartPreviousClose   float64 `json:"chartPreviousClose"`
				PreviousClose        float64 `json:"previousClose"`
				Scale                int     `json:"scale"`
				PriceHint            int     `json:"priceHint"`
				CurrentTradingPeriod struct {
					Pre struct {
						Timezone  string `json:"timezone"`
						End       int    `json:"end"`
						Start     int    `json:"start"`
						Gmtoffset int    `json:"gmtoffset"`
					} `json:"pre"`
					Regular struct {
						Timezone  string `json:"timezone"`
						End       int    `json:"end"`
						Start     int    `json:"start"`
						Gmtoffset int    `json:"gmtoffset"`
					} `json:"regular"`
					Post struct {
						Timezone  string `json:"timezone"`
						End       int    `json:"end"`
						Start     int    `json:"start"`
						Gmtoffset int    `json:"gmtoffset"`
					} `json:"post"`
				} `json:"currentTradingPeriod"`
				TradingPeriods [][]struct {
					Timezone  string `json:"timezone"`
					End       int    `json:"end"`
					Start     int    `json:"start"`
					Gmtoffset int    `json:"gmtoffset"`
				} `json:"tradingPeriods"`
				DataGranularity string   `json:"dataGranularity"`
				Range           string   `json:"range"`
				ValidRanges     []string `json:"validRanges"`
			} `json:"meta"`
			Timestamp  []int `json:"timestamp"`
			Indicators struct {
				Quote []struct {
					Low    []float64 `json:"low"`
					High   []float64 `json:"high"`
					Volume []int     `json:"volume"`
					Open   []float64 `json:"open"`
					Close  []float64 `json:"close"`
				} `json:"quote"`
			} `json:"indicators"`
		} `json:"result"`
		Error interface{} `json:"error"`
	} `json:"chart"`
}
```

We get a plethora of other information, but importantly, volume, so we have an idea of how much interest is in the stock. If the volume is low, we know there's little interest in the asset, and vice-versa if it's high (GME during the craze had very high volume compared to historically). 

I think we'd have to stand up a different service that pub-subs the updated calculations for spread on that service (we don't want to interfere with this machine trying to complete orders as fast as possible).

This would have to be pretty dynamic. If we also took into account losing orders, we might have a threshold where we want to win 1% of all *good* trades, and if we find ourselves only winning some fraction of that, we might believe that we have a spread that is too high for the market (lots of competition), narrow our spread, and try to get back to winning 1% of our trades. We could keep track of that in the service itself, but I think that'd be important data to persist in a data lake so that analysts/traders could analyze it later. 

We could also dynamically adjust based on historical prices, as you suggest. This makes sense for an optimization, but I'm stumped on how to do it fast enough, since we want to keep the logic of accepting offers fast.

One way we could do that is to shard our data by asset:

If we had 5000 tickers we cared about, instead of having one set of machines listening to all offers and trying to execute them in an idempotent fashion, we could have all machines pull from the same data source, but only execute orders on the assets they are assigned (maybe 10-100), so we'd have 50-500 clusters, and they could keep more detailed historical price data in memory to make more detailed decisions on when to buy/sell an asset.

We could use the same schema as the `transactions` as well for those losing trades, but write to a different database, because the `transactions` database is on the hotpath.

Reply:
:clap: